### ISSUE TITLE
Shadow lint shouldn't fire on extraction

### ISSUE THREAD

**Issue by taralx** [Manishearth/rust-clippy#617]
I often write this kind of code:

let x = match o { Some(Deep(Extraction(ref x))) => x, _ => return ... }

This produces a shadow_unrelated lint, but probably shouldn't.

**Comment by llogiq** [comment]
Yeah, that's a bug. I think it may be due to how `check_pat(..)` tries to find the relevant structs/fields. I may want to change this somewhat.

**Comment by llogiq** [comment]
Hmmm...adding

```
struct Deep<T>(T);
```
and in `main()`:

```
    let deep : Option<Deep<Option<u8>>> = Some(Deep(Some(1)));
    let deep = match deep {
        Some(Deep(Some(ref d))) => d,
        _ => &y,
    };
```

I get a `shadow_reuse` match. @taralx could you post a complete example we can test against?

**Comment by taralx** [comment]
Derp. The shadow_unrelated lint is actually a fair one in my code. I had this:

```rust
let d = match o { Object::LazyStream{ref d,..} => d, ... };
...
match o {
  Object::LazyStream{d,..} => Ok(d),
  ...
}
```

**Comment by llogiq** [comment]
So should I close this issue?

**Comment by taralx** [comment]
Yeah, that's cool.
